### PR TITLE
Refactor aioseop-count-chars.js

### PR DESCRIPTION
**Pull request by arnaudbroes**
Issue #2337 & #2413 & #2747 & #2755 & #2927 & #2975 & #2977

## Proposed changes

Refactors the character counter code and fixes numerous bugs (see above).

## Types of changes

- Bugfix (non-breaking change which fixes an issue)
- Improves existing functionality
- Improves existing code

## Checklist

- [x] I've selected the appropriate branch (ie x.y rather than master).
- [x] I've created an appropriate title, descriptive of what the PR does.
- [ ] Travis-ci runs with no errors.

## Testing instructions

Basically, make sure that the character counter changes whenever anything changes inside the meta box or the editor and that all of the above issues have been fixed.

### CODE REVIEW

**Comment by arnaudbroes** [comment]
@wpsmort this can be tested now.

**Comment by arnaudbroes** [comment]
Failed Travis CI build can be ignored - the fix for it is in another PR.

**Comment by wpsmort** [comment]
@arnaudbroes I have tested this but there are a number of bugs and some errors in the browser error console.  Here's what I have:

1. The character counter fields on General Settings > Home Page Settings don't change color until you place your cursor in the field.
2. There are two console errors on the General Settings screen - "Cannot read property 'length' of undefined" and "tinymce is not defined".
3. The character counter fields on Social Meta > Home Page Settings don't change color at all.
4. There's a console error "aioseopCharacterCounter is not defined" when you're on the Social Meta screen.
5. The character counter doesn't change when you type a title on a new post.
6. The character counter is doesn't include the number of characters of the post title unless Autogenerate Descriptions is enabled (most likely the same bug as above).
7. The character counter doesn't work at for autogenerated title and description on the Social Settings tab.

**Comment by wpsmort** [comment]
@arnaudbroes I tested the latest commit as requested and found some remaining bugs as follows:

1. The character counter for Social Meta > Home Page Settings > Home Title changes to yellow at 56 chars when it should be 86 chars (it's incorrectly using the same threshold as the Home Page SEO Title).

2. The character counter for Social Meta > Home Page Settings > Home Description changes to yellow at 56 chars when it should be 191 chars (the threshold for red is fine for both this a 1 above).

3. The character counters on the New Post screen don't increment when you start typing a title or content/excerpt in Classic Editor or Block Editor.  They only change wen you save the post.  Once you save, they increment fine.

4. The character counter doesn't increment at all for the SEO Description field, OG Title and OG Description fields on Edit Category.  The count is also wrong when you type in the SEO Title field, even after saving.

**Comment by arnaudbroes** [comment]
@wpsmort (2) has been reverted to the old code as discussed; (1) has been fixed and was caused by the fact that new Posts use a different page than existing Posts.

**Comment by wpsmort** [comment]
This is tested and is good.

**Comment by arnaudbroes** [comment]
@EkoJR FYI the old script is still there as we need it to support taxonomies for the time being (until we refactor the title and description code).

File: js/admin/aioseop-count-chars.js [hunk 1]
```diff
@@ -3,69 +3,356 @@
...
-var aiosp_title_extra = parseInt( aioseop_count_chars.extraTitleLength, 10 );
+if ('undefined' !== typeof aioseopCharacterCounter) {
+	var aioseopCharacterCounterIsGutenberg = aioseopCharacterCounter.isGutenberg;
+	var aioseopCharacterCounterAutogenerateDescriptions = aioseopCharacterCounter.autogenerateDescriptions;
+	var extraTitleLength = +aioseopCharacterCounter.extraTitleLength;
+	currentPage = aioseopCharacterCounter.currentPage;
+	pluginDirName = aioseopCharacterCounter.pluginDirName;
+} else if ('undefined' !== typeof aioseopOGCharacterCounter) {
+	pluginDirName = aioseopOGCharacterCounter.pluginDirName;
+	currentPage = aioseopOGCharacterCounter.currentPage;
```
Comment (EkoJR) [inline js/admin/aioseop-count-chars.js]: Variable alignment with the `=` could make this easier to read, but the long variable name `aioseopCharacterCounterAutogenerateDescriptions` makes it rather difficult to.

File: aioseop_class.php [hunk 2]
```diff
@@ -3673,42 +3673,62 @@ public function admin_enqueue_scripts( $hook_suffix ) {
...
 
-		$extra_title_len = 0;
+		$count_chars_data = array();
 		switch ( $hook_suffix ) {
+			case 'term.php':
+				// Legacy code for taxonomy terms until we refactor all title format related code.
+				$count_chars_data['aiosp_title_extra'] = 0;
+				wp_enqueue_script(
+					'aioseop-count-chars-old',
+					AIOSEOP_PLUGIN_URL . 'js/admin/aioseop-count-chars-old.min.js',
```
[the remark this answers is not in the document]
Comment (arnaudbroes) [reply aioseop_class.php]: Agreed. I already opened an issue for it, but I've now also quoted your comment there - https://github.com/semperfiwebdesign/all-in-one-seo-pack/issues/2991. In the meanwhile I've reverted the code back to unminified files.

File: js/admin/aioseop-count-chars.js (hunk 1 above)
Comment (arnaudbroes) [reply js/admin/aioseop-count-chars.js]: It does look better aligned so I've made the change. I'm going to change the variable from using `var` to `let` since I initially thought that Internet Explorer doesn't support this, but it looks like it only doesn't support it for `for` control structures.

Comment (arnaudbroes) [reply js/admin/aioseop-count-chars.js]: I've resorted to using `let` (ES6) to define them inside a block scope and shorten the variable names. It looks much better now, but as a result, `grunt-contrib-uglify` breaks because it doesn't support ES6. I've created an issue to migrate to TTerser instead (which is much popular at the moment and which does support ES6) - https://github.com/semperfiwebdesign/all-in-one-seo-pack/issues/3001. I'll make this change when I also tackle the issue to use minified JS files.

File: js/admin/aioseop-count-chars.js [hunk 3]
```diff
@@ -3,69 +3,356 @@
...
- * @return Mixed.
- */
-function aioseopCountChars( field, cntfield ) {
-	var extra = 0;
-	var field_size;
-	if ( ( field.attr('name') === 'aiosp_title' ) && ( typeof aiosp_title_extra !== 'undefined' ) ) {
-		extra = aiosp_title_extra;
+	aioseopAddEventListeners();
+	// Fire once on page load.
+	aioseopCountChars();
```
[the remark this answers is not in the document]
Comment (arnaudbroes) [reply js/admin/aioseop-count-chars.js]: I've moved everything inside the $(function() {} ); statement. Had to add a small comment for eslint to ignore the amount of statements inside since it's just a wrapper.

**Comment by arnaudbroes** [comment]
This can be reviewed again. @EkoJR definitely check out my comments.

File: aioseop_class.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (EkoJR) [reply aioseop_class.php]: I've looked at a couple plugins that do minify the admin side, and they all were using a debug constant. Seeing how this is new code being introduced, it seemed wise to have a debug mode added first.

I do support the idea of using minify, but it does seems like majority of the public reports could change to line errors from a minified files.

File: js/admin/aioseop-count-chars.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (EkoJR) [reply js/admin/aioseop-count-chars.js]: Oh, an anonymous function...yea, that would be done differently than what I commented.

The error you are getting is from the number of Variable Statements you have wrapped in that single function. I believe the limit is 10. The only recommendation I would have is to reduce the number of variables, but I'm ok with ignoring that report.

[the remark this answers is not in the document]
Comment (arnaudbroes) [reply js/admin/aioseop-count-chars.js]: It's not the amount of variables, but the amount of statements inside that triggers the warning.  The max. is 30 and there are about 35 at the moment, but since it's a wrapper we shouldn't be strict about following the coding standards. I had a quick look around online and looks like other people do the same thing, so good to go!
